File: packages/capacitor/android/mushi-android/src/main/kotlin/dev/mushimushi/sdk/capture/ProactiveDetector.kt
package dev.mushimushi.sdk.capture

import android.view.Choreographer
import android.view.MotionEvent
import android.view.View

/**
 * Proactive trigger detection for Android. Mirrors `proactive-triggers.ts`
 * from the web SDK adapted for Android touch and frame budget:
 *
 * - **Rage-tap**: ≥3 `ACTION_DOWN` events on the same view within 500 ms.
 * - **Slow-screen**: a vsync frame that takes >200 ms, detected via
 *   [Choreographer.FrameCallback].
 *
 * Usage: call [install] once from [dev.mushimushi.sdk.Mushi.init], passing
 * the root [View] (e.g. `activity.window.decorView`).
 */
class ProactiveDetector(private val config: Config = Config()) {

    data class Config(
        /** Enable rage-tap detection (≥3 taps on same view in 500 ms). */
        val rageTap: Boolean = true,
        /** Enable slow-screen detection (frame >200 ms). */
        val slowScreen: Boolean = true,
        /** Threshold in ms to flag a frame as slow. */
        val slowScreenThresholdMs: Long = 200L,
        /** Max proactive triggers fired per session. */
        val maxPerSession: Int = 3,
    )

    fun interface TriggerCallback {
        fun onTrigger(type: String, context: Map<String, Any>)
    }

    private var callback: TriggerCallback? = null
    private var fired = 0

    // Rage-tap state
    private val tapTimes = mutableListOf<Long>()
    @Volatile private var lastTapView: View? = null

    // Slow-screen state
    private var choreographer: Choreographer? = null
    private var lastFrameNano: Long = 0L
    private var frameCallbackInstalled = false

    private val frameCallback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (fired < config.maxPerSession) {
                if (lastFrameNano > 0L) {
                    val deltaMs = (frameTimeNanos - lastFrameNano) / 1_000_000L
                    if (deltaMs > config.slowScreenThresholdMs) {
                        fire("slow_screen", mapOf("frameMs" to deltaMs))
                    }
                }
                lastFrameNano = frameTimeNanos
            }
            choreographer?.postFrameCallback(this)
        }
    }

    /**
     * Install on the root [View]. The [view]'s [View.OnTouchListener] is used
     * for rage-tap detection; an existing listener is chained rather than replaced.
     *
     * Must be called on the main thread.
     */
    fun install(view: View, callback: TriggerCallback) {
        this.callback = callback
        if (config.rageTap) {
            val existing = view.getTag(TAG_PREV_LISTENER) as? View.OnTouchListener
            view.setOnTouchListener { v, event ->
                onTouch(v, event)
                existing?.onTouch(v, event) ?: false
            }
            view.setTag(TAG_PREV_LISTENER, existing)
        }
        if (config.slowScreen && !frameCallbackInstalled) {
            choreographer = Choreographer.getInstance()
            choreographer?.postFrameCallback(frameCallback)
            frameCallbackInstalled = true
        }
    }

    fun destroy() {
        callback = null
        choreographer = null
        frameCallbackInstalled = false
    }

    private fun onTouch(view: View, event: MotionEvent) {
        if (event.action != MotionEvent.ACTION_DOWN) return
        if (fired >= config.maxPerSession) return
        val now = System.currentTimeMillis()
        if (view === lastTapView) {
            tapTimes.add(now)
            tapTimes.removeAll { now - it > 500L }
            if (tapTimes.size >= 3) {
                fire("rage_tap", mapOf(
                    "tapCount" to tapTimes.size,
                    "viewClass" to view.javaClass.simpleName,
                    "contentDescription" to (view.contentDescription?.toString() ?: ""),
                ))
                tapTimes.clear()
            }
        } else {
            lastTapView = view
            tapTimes.clear()
            tapTimes.add(now)
        }
    }

    private fun fire(type: String, context: Map<String, Any>) {
        fired++
        callback?.onTrigger(type, context)
    }

    private companion object {
        val TAG_PREV_LISTENER = View.generateViewId()
    }
}

File: packages/ios/Sources/MushiMushi/Capture/ProactiveDetector.swift
#if os(iOS)
import UIKit
import QuartzCore

/// Proactive trigger detection for iOS. Mirrors `proactive-triggers.ts`
/// from the web SDK adapted for UIKit/SwiftUI:
///
/// - **Rage-tap**: ≥3 taps on the same view within 500 ms.
/// - **Slow-screen**: a main-thread frame that takes >200 ms to render,
///   detected via a `CADisplayLink` monitoring the vsync budget.
///
/// Call `install(in:onTrigger:)` once in `Mushi.configure()`. The detector
/// holds a weak reference to the root window so it doesn't keep the scene
/// alive.
public final class ProactiveDetector: NSObject {

    public struct Config {
        /// Enable rage-tap detection (≥3 taps on same view in 500 ms).
        public var rageTap: Bool
        /// Enable slow-screen detection (frame render >200 ms).
        public var slowScreen: Bool
        /// Threshold in ms to flag a frame as slow. Default 200.
        public var slowScreenThresholdMs: Double
        /// Maximum proactive triggers to fire per session before silencing.
        public var maxPerSession: Int

        public init(
            rageTap: Bool = true,
            slowScreen: Bool = true,
            slowScreenThresholdMs: Double = 200,
            maxPerSession: Int = 3
        ) {
            self.rageTap = rageTap
            self.slowScreen = slowScreen
            self.slowScreenThresholdMs = slowScreenThresholdMs
            self.maxPerSession = maxPerSession
        }
    }

    public typealias TriggerCallback = (_ type: String, _ context: [String: Any]) -> Void

    private let config: Config
    private var onTrigger: TriggerCallback?
    private var fired = 0

    // Rage-tap state
    private var tapTimes: [TimeInterval] = []
    private var lastTapView: UIView?
    private var tapGR: UITapGestureRecognizer?

    // Slow-screen state
    private var displayLink: CADisplayLink?
    private var lastTimestamp: CFTimeInterval = 0

    public init(config: Config = Config()) {
        self.config = config
    }

    /// Install the detector on a `UIWindow`. Safe to call from any thread;
    /// UIKit mutations are dispatched to the main queue.
    public func install(in window: UIWindow, onTrigger: @escaping TriggerCallback) {
        self.onTrigger = onTrigger

        DispatchQueue.main.async { [weak self, weak window] in
            guard let self, let window else { return }
            if self.config.rageTap {
                let gr = UITapGestureRecognizer(target: self, action: #selector(self.handleTap(_:)))
                gr.cancelsTouchesInView = false
                window.addGestureRecognizer(gr)
                self.tapGR = gr
            }
            if self.config.slowScreen {
                let dl = CADisplayLink(target: self, selector: #selector(self.displayLinkFired(_:)))
                dl.add(to: .main, forMode: .common)
                self.displayLink = dl
            }
        }
    }

    public func destroy() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if let gr = self.tapGR {
                gr.view?.removeGestureRecognizer(gr)
                self.tapGR = nil
            }
            self.displayLink?.invalidate()
            self.displayLink = nil
        }
    }

    // MARK: - Rage-tap

    @objc private func handleTap(_ gr: UITapGestureRecognizer) {
        guard fired < config.maxPerSession else { return }
        let point = gr.location(in: gr.view)
        let tapped = gr.view?.hitTest(point, with: nil)

        let now = Date().timeIntervalSince1970 * 1000
        if tapped === lastTapView {
            tapTimes.append(now)
            tapTimes = tapTimes.filter { now - $0 < 500 }
            if tapTimes.count >= 3 {
                fire("rage_tap", context: [
                    "tapCount": tapTimes.count,
                    "viewClass": String(describing: type(of: tapped as AnyObject)),
                    "accessibilityLabel": tapped?.accessibilityLabel ?? "",
                ])
                tapTimes = []
            }
        } else {
            lastTapView = tapped
            tapTimes = [now]
        }
    }

    // MARK: - Slow screen

    @objc private func displayLinkFired(_ link: CADisplayLink) {
        guard fired < config.maxPerSession else { return }
        let ts = link.timestamp
        if lastTimestamp > 0 {
            let deltaMs = (ts - lastTimestamp) * 1000
            if deltaMs > config.slowScreenThresholdMs {
                fire("slow_screen", context: ["frameMs": Int(deltaMs)])
            }
        }
        lastTimestamp = ts
    }

    private func fire(_ type: String, context: [String: Any]) {
        fired += 1
        onTrigger?(type, context)
    }
}
#endif
